presto-orc/src/main/java/com/facebook/presto/orc/writer/MapFlatColumnWriter.java

```diff
@@ -15,148 +15,568 @@
+    private int[] valueWritersLastRow = new int[0];
+
+    // If no value has been written, then valueWriters list will be empty, and we
+    // won't be able to get required column statistics for all sub nodes.
+    // This field contains cached stripe column statistics for the value node(s)
+    // for this edge case.
+    // TODO: Extend the ColumnStatisticsBuilders to create builders to all ORC
+    //  types to able to build empty stats without initializing value writers
+    private Map<Integer, ColumnStatistics> emptyValueColumnStripeStatistics;
+
+    public MapFlatColumnWriter(
+            int nodeIndex,
+            int keyNodeIndex,
+            int valueNodeIndex,
+            Type keyType,
+            Type valueType,
+            Supplier<StatisticsBuilder> keyStatisticsBuilderSupplier,
+            ColumnWriterOptions columnWriterOptions,
+            Optional<DwrfDataEncryptor> dwrfEncryptor,
+            MetadataWriter metadataWriter,
+            IntFunction<ColumnWriter> valueWriterFactory)
     {
-        // TODO: Flat map needs to get factories for key and value writers, instead of already initialized writers
-        checkArgument(column >= 0, "column is negative");
-        requireNonNull(columnWriterOptions, "columnWriterOptions is null");
-        this.column = column;
+        checkArgument(nodeIndex > 0, "nodeIndex is invalid: %s", nodeIndex);
+        checkArgument(keyNodeIndex > 0, "keyNodeIndex is invalid: %s", keyNodeIndex);
+        checkArgument(valueNodeIndex > 0, "valueNodeIndex is invalid: %s", valueNodeIndex);
+        requireNonNull(keyStatisticsBuilderSupplier, "keyStatisticsBuilderSupplier is null");
+
+        this.nodeIndex = nodeIndex;
+        this.keyNodeIndex = keyNodeIndex;
+        this.valueNodeIndex = valueNodeIndex;
+        this.keyType = requireNonNull(keyType, "keyType is null");
+        this.nullValueBlock = createNullValueBlock(requireNonNull(valueType, "valueType is null"));
+
+        this.columnWriterOptions = requireNonNull(columnWriterOptions, "columnWriterOptions is null");
+        this.dwrfEncryptor = requireNonNull(dwrfEncryptor, "dwrfEncryptor is null");
+        this.keyManager = getKeyManager(keyType, keyStatisticsBuilderSupplier);
+        this.valueWriterFactory = requireNonNull(valueWriterFactory, "valueWriterFactory is null");
+
         this.compressed = columnWriterOptions.getCompressionKind() != NONE;
-        this.columnEncoding = new ColumnEncoding(DWRF_MAP_FLAT, 0); // TODO Use c'tor with additionalSequenceEncodings
         this.metadataWriter = new CompressedMetadataWriter(metadataWriter, columnWriterOptions, dwrfEncryptor);
+        this.presentStream = new PresentOutputStream(columnWriterOptions, dwrfEncryptor);
     }
 
     @Override
     public List<ColumnWriter> getNestedColumnWriters()
     {
-        // TODO: implement me
-        return ImmutableList.of();
+        ImmutableList.Builder<ColumnWriter> nestedWriters = ImmutableList.builder();
```

Comment:
Create ImmutableList builder with size.

Reply:
Sure. However it will only work for primitive value writers.